src/java/org/apache/cassandra/index/sai/disk/v1/LongArray.java

```diff
@@ -0,0 +1,127 @@
+import java.util.function.Supplier;
+
+import org.apache.cassandra.index.sai.SSTableQueryContext;
+import org.apache.cassandra.index.sai.disk.v1.bitpack.BlockPackedReader;
+
+/**
+ * Abstraction over a long-indexed array of longs.
+ */
+public interface LongArray extends Closeable
+{
+    /**
+     * Get value at {@code idx}.
+     */
+    long get(long idx);
+
+    /**
+     * Get array length.
+     */
+    long length();
+
+    /**
+     * Using the given token returns the first row ID corresponding to the token.
+     * @param targetToken Token to lookup and it must not be smaller than previous value
+     * @return The row ID of the given token or negative value if target token is greater than all tokens
+     */
+    long findTokenRowID(long targetToken);
+
+    @Override
+    default void close() throws IOException { }
+
+    class DeferredLongArray implements LongArray
+    {
+        private final Supplier<LongArray> supplier;
+        private LongArray longArray;
+        private boolean opened = false;
+
+        public DeferredLongArray(Supplier<LongArray> supplier)
+        {
+            this.supplier = supplier;
+        }
+
+        @Override
+        public long get(long idx)
+        {
+            open();
+            return longArray.get(idx);
+        }
+
+        @Override
+        public long length()
+        {
+            open();
+            return longArray.length();
+        }
+
+        @Override
+        public long findTokenRowID(long targetToken)
+        {
+            open();
+            return longArray.findTokenRowID(targetToken);
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+            if (opened)
+                longArray.close();
+        }
+
+        private void open()
+        {
+            if (!opened)
+            {
+                longArray = supplier.get();
+                opened = true;
+            }
+        }
+    }
+
+    interface Factory
+    {
+        LongArray open();
+
+        /**
+         * TODO use a different interface for {@link BlockPackedReader}, as {@link LongArray#findTokenRowID(long)} is
+         * not supported by other implementation.
+         *
+         * @param startingIndex minimum index to be used in {@link LongArray#findTokenRowID(long)}.
+         *                      In {@link org.apache.cassandra.index.sai.disk.PostingListRangeIterator}, a segmentRowId
+         *                      is provided and then in {@link OffsetFactory},
+         *                      segment offset is applied to segmentRowId to create sstableRowId which will be used by
+         *                      {@link BlockPackedReader#openTokenReader(long, SSTableQueryContext)}.
+         * @param context shared between indexed columns for the same sstable in a given query
+         * @return token BlockPackedReader
+         */
+        default LongArray openTokenReader(long startingIndex, SSTableQueryContext context)
```

Comment:
nit: I guess it's even mentioned in the comment above, but it would be ideal to avoid "token" concepts in the interface. We get awkward lines like...

```
this.blockOffsets = new LongArray.DeferredLongArray(() -> blockOffsetsFactory.openTokenReader(0, context));
```
i.e. "Open a token reader to read block offsets?"